Persistence bundle startup failure does not cause Stargate process failure

Currently if a persistence bundle fails to start with an exception, the overall Stargate process remains running (see below for a DSE example).

Since persistence is a fundamental service of Stargate, it does not make sense to keep the process running as it will not be usable and will not re-attempt connecting to the persistence backend in this state.

```
ERROR [DSE main thread] 2021-01-12 14:34:34,314 CassandraDaemon.java:821 - Exception encountered during startup
java.lang.RuntimeException: A node with address /127.0.2.1 already exists, cancelling join. Use cassandra.replace_address if you want to replace this node.
	at org.apache.cassandra.service.StorageService.checkForEndpointCollision(StorageService.java:786)
	at org.apache.cassandra.service.StorageService.prepareToJoin(StorageService.java:1157)
	at org.apache.cassandra.service.StorageService.initServer(StorageService.java:918)
	at org.apache.cassandra.service.StorageService.initServer(StorageService.java:852)
	at org.apache.cassandra.service.CassandraDaemon.setup(CassandraDaemon.java:419)
	at org.apache.cassandra.service.CassandraDaemon.activate0(CassandraDaemon.java:754)
	at org.apache.cassandra.service.CassandraDaemon.access$100(CassandraDaemon.java:88)
	at org.apache.cassandra.service.CassandraDaemon$3.run(CassandraDaemon.java:715)
java.lang.RuntimeException: A node with address /127.0.2.1 already exists, cancelling join. Use cassandra.replace_address if you want to replace this node.
	at org.apache.cassandra.service.StorageService.checkForEndpointCollision(StorageService.java:786)
	at org.apache.cassandra.service.StorageService.prepareToJoin(StorageService.java:1157)
	at org.apache.cassandra.service.StorageService.initServer(StorageService.java:918)
	at org.apache.cassandra.service.StorageService.initServer(StorageService.java:852)
	at org.apache.cassandra.service.CassandraDaemon.setup(CassandraDaemon.java:419)
	at org.apache.cassandra.service.CassandraDaemon.activate0(CassandraDaemon.java:754)
	at org.apache.cassandra.service.CassandraDaemon.access$100(CassandraDaemon.java:88)
	at org.apache.cassandra.service.CassandraDaemon$3.run(CassandraDaemon.java:715)
ERROR: Bundle io.stargate.db.dse [1] EventDispatcher: Error during dispatch. (java.lang.RuntimeException: Unable to start DSE persistence layer)
java.lang.RuntimeException: Unable to start DSE persistence layer
	at io.stargate.db.dse.impl.DsePersistence.initializePersistence(DsePersistence.java:188)
	at io.stargate.db.dse.impl.DsePersistence.initializePersistence(DsePersistence.java:85)
	at io.stargate.db.datastore.common.AbstractCassandraPersistence.initialize(AbstractCassandraPersistence.java:106)
	at io.stargate.db.dse.DsePersistenceActivator.createService(DsePersistenceActivator.java:142)
	at io.stargate.core.activator.BaseActivator.createServices(BaseActivator.java:187)
	at io.stargate.core.activator.BaseActivator.startServiceInternal(BaseActivator.java:123)
	at io.stargate.core.activator.BaseActivator.access$700(BaseActivator.java:34)
	at io.stargate.core.activator.BaseActivator$Tracker.startIfAllRegistered(BaseActivator.java:174)
	at io.stargate.core.activator.BaseActivator$Tracker.addingService(BaseActivator.java:150)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:943)
	at org.osgi.util.tracker.ServiceTracker$Tracked.customizerAdding(ServiceTracker.java:871)
	at org.osgi.util.tracker.AbstractTracked.trackAdding(AbstractTracked.java:256)
	at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:229)
	at org.osgi.util.tracker.ServiceTracker$Tracked.serviceChanged(ServiceTracker.java:903)
	at org.apache.felix.framework.EventDispatcher.invokeServiceListenerCallback(EventDispatcher.java:990)
	at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:838)
	at org.apache.felix.framework.EventDispatcher.fireServiceEvent(EventDispatcher.java:545)
	at org.apache.felix.framework.Felix.fireServiceEvent(Felix.java:4833)
	at org.apache.felix.framework.Felix.registerService(Felix.java:3804)
	at org.apache.felix.framework.BundleContextImpl.registerService(BundleContextImpl.java:328)
	at org.apache.felix.framework.BundleContextImpl.registerService(BundleContextImpl.java:302)
	at io.stargate.core.activator.BaseActivator.startServiceInternal(BaseActivator.java:128)
	at io.stargate.core.activator.BaseActivator.start(BaseActivator.java:67)
	at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:698)
	at org.apache.felix.framework.Felix.activateBundle(Felix.java:2402)
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2308)
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998)
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:984)
	at io.stargate.starter.Starter.start(Starter.java:426)
	at io.stargate.starter.Starter.cli(Starter.java:597)
	at io.stargate.starter.Starter.main(Starter.java:638)
Caused by: java.lang.RuntimeException: Exception encountered during startup
	at org.apache.cassandra.service.CassandraDaemon.exitOrFail(CassandraDaemon.java:927)
	at org.apache.cassandra.service.CassandraDaemon.activate0(CassandraDaemon.java:824)
	at org.apache.cassandra.service.CassandraDaemon.access$100(CassandraDaemon.java:88)
	at org.apache.cassandra.service.CassandraDaemon$3.run(CassandraDaemon.java:715)
Caused by: java.lang.RuntimeException: A node with address /127.0.2.1 already exists, cancelling join. Use cassandra.replace_address if you want to replace this node.
	at org.apache.cassandra.service.StorageService.checkForEndpointCollision(StorageService.java:786)
	at org.apache.cassandra.service.StorageService.prepareToJoin(StorageService.java:1157)
	at org.apache.cassandra.service.StorageService.initServer(StorageService.java:918)
	at org.apache.cassandra.service.StorageService.initServer(StorageService.java:852)
	at org.apache.cassandra.service.CassandraDaemon.setup(CassandraDaemon.java:419)
	at org.apache.cassandra.service.CassandraDaemon.activate0(CassandraDaemon.java:754)
	... 2 more
INFO  [main] 2021-01-12 14:37:48,468 BaseActivator.java:126 - Registering core services as com.codahale.metrics.health.HealthCheckRegistry
INFO  [main] 2021-01-12 14:37:48,468 BaseActivator.java:132 - Started core services
Starting bundle io.stargate.cql
INFO  [main] 2021-01-12 14:37:48,797 BaseActivator.java:64 - Starting CQL ...
Starting bundle io.stargate.graphql
INFO  [main] 2021-01-12 14:37:48,973 BaseActivator.java:64 - Starting GraphQL ...
Starting bundle io.stargate.health
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [bundle://a49c8615-71bc-4346-8a36-8b7890b6c76a_20.0:82/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [bundle://a49c8615-71bc-4346-8a36-8b7890b6c76a_20.0:100/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [ch.qos.logback.classic.util.ContextSelectorStaticBinder]
INFO  [main] 2021-01-12 14:37:49,161 BaseActivator.java:64 - Starting healthchecker ...
INFO  [main] 2021-01-12 14:37:49,162 HealthCheckerActivator.java:30 - Starting healthchecker....
INFO  [2021-01-12 19:37:49,929] io.dropwizard.server.SimpleServerFactory: Registering jersey handler with root path prefix: /checker/
WARN  [2021-01-12 19:37:49,929] org.eclipse.jetty.server.handler.ContextHandler: i.d.j.MutableServletContextHandler@508b4f70{/,null,UNAVAILABLE} contextPath ends with /
INFO  [2021-01-12 19:37:49,931] io.dropwizard.server.SimpleServerFactory: Registering admin handler with root path prefix: /
INFO  [2021-01-12 19:37:49,937] io.dropwizard.server.SimpleServerFactory: Registering jersey handler with root path prefix: /checker/
WARN  [2021-01-12 19:37:49,937] org.eclipse.jetty.server.handler.ContextHandler: i.d.j.MutableServletContextHandler@508b4f70{/checker,null,UNAVAILABLE} contextPath ends with /
INFO  [2021-01-12 19:37:49,937] io.dropwizard.server.SimpleServerFactory: Registering admin handler with root path prefix: /
INFO  [2021-01-12 19:37:49,937] io.dropwizard.server.ServerFactory: Starting Server
INFO  [2021-01-12 19:37:50,011] org.eclipse.jetty.setuid.SetUIDListener: Opened Server@7e43b342{HTTP/1.1, (http/1.1)}{127.0.2.1:8084}
INFO  [2021-01-12 19:37:50,013] org.eclipse.jetty.server.Server: jetty-9.4.30.v20200611; built: 2020-06-11T12:34:51.929Z; git: 271836e4c1f4612f12b7bb13ef5a92a927634b0d; jvm 1.8.0_172-b11
INFO  [2021-01-12 19:37:50,429] io.dropwizard.jersey.DropwizardResourceConfig: The following paths were found for the configured resources:

    GET     /liveness (io.stargate.health.CheckerResource)
    GET     /readiness (io.stargate.health.CheckerResource)

INFO  [2021-01-12 19:37:50,430] org.eclipse.jetty.server.handler.ContextHandler: Started i.d.j.MutableServletContextHandler@508b4f70{/checker,null,AVAILABLE}
INFO  [2021-01-12 19:37:50,432] io.dropwizard.setup.AdminEnvironment: tasks = 

    POST    /tasks/log-level (io.dropwizard.servlets.tasks.LogConfigurationTask)
    POST    /tasks/gc (io.dropwizard.servlets.tasks.GarbageCollectionTask)

WARN  [2021-01-12 19:37:50,432] io.dropwizard.setup.AdminEnvironment: 
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!    THIS APPLICATION HAS NO HEALTHCHECKS. THIS MEANS YOU WILL NEVER KNOW      !
!     IF IT DIES IN PRODUCTION, WHICH MEANS YOU WILL NEVER KNOW IF YOU'RE      !
!    LETTING YOUR USERS DOWN. YOU SHOULD ADD A HEALTHCHECK FOR EACH OF YOUR    !
!         APPLICATION'S DEPENDENCIES WHICH FULLY (BUT LIGHTLY) TESTS IT.       !
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
INFO  [2021-01-12 19:37:50,437] org.eclipse.jetty.server.handler.ContextHandler: Started i.d.j.MutableServletContextHandler@238280df{/,null,AVAILABLE}
INFO  [2021-01-12 19:37:50,449] org.eclipse.jetty.server.AbstractConnector: Started Server@7e43b342{HTTP/1.1, (http/1.1)}{127.0.2.1:8084}
INFO  [2021-01-12 19:37:50,449] org.eclipse.jetty.server.Server: Started @207953ms
INFO  [2021-01-12 19:37:50,449] io.stargate.health.HealthCheckerActivator: Started healthchecker....
INFO  [main] 2021-01-12 14:37:50,449 BaseActivator.java:132 - Started healthchecker
Starting bundle org.apache.felix.scr
Starting bundle null
Starting bundle null
Starting bundle osgi.cmpn
Starting bundle org.osgi.util.function
Starting bundle org.osgi.util.promise
Starting bundle io.stargate.db
INFO  [main] 2021-01-12 14:37:50,474 BaseActivator.java:64 - Starting DB services ...
Starting bundle io.stargate.web
INFO  [main] 2021-01-12 14:37:50,687 BaseActivator.java:64 - Starting restapi ...
Finished starting bundles.
```

--- a/core/src/main/java/io/stargate/core/activator/BaseActivator.java
+++ b/core/src/main/java/io/stargate/core/activator/BaseActivator.java
@@ -120,16 +120,21 @@ public abstract class BaseActivator implements BundleActivator {
       return;
     }
     started = true;
-    List<ServiceAndProperties> services = createServices();
-    for (ServiceAndProperties service : services) {
-      if (service != null) {
-        logger.info("Registering {} as {}", activatorName, service.targetServiceClass.getName());
-        targetServiceRegistrations.add(
-            context.registerService(
-                service.targetServiceClass.getName(), service.service, service.properties));
+
+    try {
+      List<ServiceAndProperties> services = createServices();
+      for (ServiceAndProperties service : services) {
+        if (service != null) {
+          logger.info("Registering {} as {}", activatorName, service.targetServiceClass.getName());
+          targetServiceRegistrations.add(
+              context.registerService(
+                  service.targetServiceClass.getName(), service.service, service.properties));
+        }
       }
+      logger.info("Started {}", activatorName);
+    } catch (Exception e) {
+      throw new ServiceStartException("Unable to start " + activatorName, e);
     }
-    logger.info("Started {}", activatorName);
   }
 
   public class Tracker extends ServiceTracker<Object, Object> {

--- a/core/src/main/java/io/stargate/core/activator/ServiceStartException.java
+++ b/core/src/main/java/io/stargate/core/activator/ServiceStartException.java
@@ -1,0 +1,24 @@
+/*
+ * Copyright The Stargate Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.core.activator;
+
+/** This exception is used to communicate critical service startup errors to the Starter. */
+public class ServiceStartException extends RuntimeException {
+
+  public ServiceStartException(String message, Throwable cause) {
+    super(message, cause);
+  }
+}

--- a/stargate-starter/src/main/java/io/stargate/starter/Starter.java
+++ b/stargate-starter/src/main/java/io/stargate/starter/Starter.java
@@ -54,6 +54,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Optional;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import javax.inject.Inject;
 import org.apache.commons.collections4.iterators.PeekingIterator;
 import org.apache.commons.lang3.math.NumberUtils;
@@ -65,6 +66,8 @@ import org.osgi.framework.Bundle;
 import org.osgi.framework.BundleContext;
 import org.osgi.framework.BundleException;
 import org.osgi.framework.Constants;
+import org.osgi.framework.FrameworkEvent;
+import org.osgi.framework.FrameworkListener;
 import org.osgi.framework.InvalidSyntaxException;
 import org.osgi.framework.ServiceReference;
 
@@ -249,6 +252,7 @@ public class Starter {
   private Felix framework;
   private List<Bundle> bundleList;
   private boolean watchBundles = true;
+  private AtomicBoolean startError = new AtomicBoolean();
 
   public Starter() {}
 
@@ -390,6 +394,7 @@ public class Starter {
 
     // Install bundles
     context = framework.getBundleContext();
+    context.addFrameworkListener(new BundleFailureListener());
     File[] files = new File(JAR_DIRECTORY).listFiles();
     List<File> jars = pickBundles(files);
     framework.start();
@@ -424,6 +429,11 @@ public class Starter {
     for (Bundle bundle : bundleList) {
       System.out.println("Starting bundle " + bundle.getSymbolicName());
       bundle.start();
+    }
+
+    if (startError.get()) {
+      System.out.println("Terminating due to previous service startup errors.");
+      System.exit(1);
     }
 
     System.out.println(STARTED_MESSAGE);
@@ -636,6 +646,29 @@ public class Starter {
 
   public static void main(String[] args) {
     cli(args, Starter.class);
+  }
+
+  private class BundleFailureListener implements FrameworkListener {
+
+    @Override
+    public void frameworkEvent(FrameworkEvent event) {
+      Throwable throwable = event.getThrowable();
+      if (throwable != null) {
+        // We rely on the exception class name here because there is no easy way to share
+        // classes between the Starter and bundles, and this use case does not seem worth
+        // adding a new OSGi artifact.
+        if (throwable.getClass().getSimpleName().equals("ServiceStartException")) {
+          startError.set(true);
+
+          System.out.printf(
+              "Detected service startup failure in bundle %s: %s%n",
+              event.getBundle().getSymbolicName(), throwable);
+
+          // This should be logged by OSGi, but just in case dump it to STDERR too
+          throwable.printStackTrace(System.err);
+        }
+      }
+    }
   }
 
   public static class NodeToolOptionParser<T> extends AbstractOptionParser<T> {

